pipeline_steps/gate.groovy

```diff
@@ -0,0 +1,194 @@
+                    parameters: [
+                        [
+                            $class: "StringParameterValue",
+                            name: "RPC_GATING_BRANCH",
+                            value: RPC_GATING_BRANCH,
+                        ],
+                        [
+                            $class: "StringParameterValue",
+                            name: "BRANCH",
+                            value: sha1,
+                        ]
+                    ]
+                )
+            }
+        }
+    }
+
+    parallel parallelBuilds
+
+    build(
+      job: "Merge-Pull-Request",
+      wait: false,
+      parameters: [
+        [
+          $class: "StringParameterValue",
+          name: "RPC_GATING_BRANCH",
+          value: RPC_GATING_BRANCH,
+        ],
+        [
+          $class: "StringParameterValue",
+          name: "pr_repo",
+          value: ghprbGhRepository,
+        ],
+        [
+          $class: "StringParameterValue",
+          name: "pr_number",
+          value: ghprbPullId,
+        ],
+        [
+          $class: "StringParameterValue",
+          name: "commit",
+          value: ghprbActualCommit,
+        ],
+      ]
+    )
+}
+
+
+void testPullRequestWithAntecedents(String repo_name){
+    allGateJobNames = Hudson.instance.getAllItems(org.jenkinsci.plugins.workflow.job.WorkflowJob)*.fullName
+    gateJobNames = (allGateJobNames.findAll {it =~ /GATE_${repo_name}/}).sort(false)
+    println("Discovered the following pull request gate jobs for repo ${repo_name}:\n${gateJobNames.join("\n")}")
+
+    triggerBuild = currentBuild.rawBuild
+    triggerJobName = triggerBuild.getParent().displayName
+    triggerJob = Hudson.instance.getItems(org.jenkinsci.plugins.workflow.job.WorkflowJob).find {it.displayName == triggerJobName}
+    antecedentTriggerBuilds = []
+    success = Result.fromString("SUCCESS")
+    pullRequestIDs = []
+    gateBuilds = []
+
+    while (true) {
+        // Check antecedents for failures
+        currentAntecedentTriggerBuilds = triggerJob.getBuilds().findAll {
+            (
+                it.isBuilding() == true
+                && it.getCauses()[0] instanceof org.jenkinsci.plugins.ghprb.GhprbCause
+                && it.getNumber() < triggerBuild.getNumber()
+            )
+        }.reverse()
+        if (currentAntecedentTriggerBuilds){
+            println("Active antecedent builds:\n${currentAntecedentTriggerBuilds.join("\n")}")
+        }else {
+            println("No active antecedent builds.")
+        }
+        if (antecedentTriggerBuilds.collect(){it.getNumber()} != currentAntecedentTriggerBuilds.collect(){it.getNumber()}) {
+            failed = antecedentTriggerBuilds.findAll() {it.getResult() && it.getResult().isWorseThan(success)}
+            if (failed) {
+                println("Antecedent builds that have completed with failure:\n${failed.join("\n")}")
+                println("Killing existing builds that depend on failed antecedents.")
```

Comment:
I'll take a look at this. I'm tempted to say it should be a follow up task because end users shouldn't really need to look at these builds. This code lists the cause of the failure, i.e. the downstream builds that failed, at the end of the log file so I think it's clear what's happening.

Reply:
Its clear in the trigger job, but not in the job that gets killed, and the trigger is not necessarily an obvious place to look for the cause if you're not familiar with the structure of related jobs.

Reply:
The pull request only links to the trigger build so the only way the user finds the downstream build is via the trigger build log, well unless they are searching through all the builds on particular jobs but that would be pretty silly. The trigger build prints failures at the end of the log so there is a logical sequence from pull request to any failures. Again I'm not dismissing the change but a standard user workflow in terms of reviewing any failure should not require them to view builds killed by the trigger build.